File: Editorial/1940_Longest_Common_Subsequence_Between_Sorted_Arrays/Python3/3_Binary_Search.py
class Solution:
    def longestCommonSubsequence(self, arrays: List[List[int]]) -> List[int]:

        def binary_search(target, nums):
            left = 0
            right = len(nums) - 1
            while left <= right:
                mid = left + (right - left) // 2
                if nums[mid] > target:
                    right = mid - 1
                elif nums[mid] < target:
                    left = mid + 1
                else:
                    return True
            return False

        shortestArray = arrays[0]
        for array in arrays:
            if len(array) < len(shortestArray):
                shortestArray = array

        longest_common_subseq = shortestArray

        for array in arrays:
            # There are no elements that are common to all of the arrays
            if len(longest_common_subseq) == 0:
                return longest_common_subseq
            uncommon = []

            # Remove any elements from the longest common subsequence
            # that are not in current array
            for num in longest_common_subseq:
                if not binary_search(num, array):
                    uncommon.append(num)
            for num in uncommon:
                longest_common_subseq.remove(num)

        return longest_common_subseq

File: Editorial/1940_Longest_Common_Subsequence_Between_Sorted_Arrays/CPP/3_Binary_Search.cpp
class Solution {
public:
    vector<int> longestCommonSubsequence(vector<vector<int>>& arrays) {
        vector<int> shortestArray = arrays[0];
        for (const auto& array : arrays) {
            if (array.size() < shortestArray.size()) {
                shortestArray = array;
            }
        }

        vector<int> longestCommonSubseq(shortestArray.begin(),
                                        shortestArray.end());

        for (const auto& array : arrays) {
            // There are no elements that are common to all of the arrays
            if (longestCommonSubseq.empty()) {
                return longestCommonSubseq;
            }

            // Remove any elements from the longest common subsequence
            // that are not in current array
            vector<int> uncommon;
            for (int num : longestCommonSubseq) {
                if (!binarySearch(num, array)) {
                    uncommon.push_back(num);
                }
            }
            for (int num : uncommon) {
                longestCommonSubseq.erase(
                    remove(longestCommonSubseq.begin(),
                           longestCommonSubseq.end(), num),
                    longestCommonSubseq.end());
            }
        }

        return longestCommonSubseq;
    }

private:
    bool binarySearch(int target, const vector<int>& nums) {
        int left = 0;
        int right = nums.size() - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] > target) {
                right = mid - 1;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                return true;
            }
        }
        return false;
    }
};
